File: ID10196_CheckTheCheck.py
# @JUDGE_ID: 10196 Python "Check the Check" 

# @BEGIN_OF_SOURCE_CODE

from sys import stdin

BOARD_LEN = 8

def EmptyBoard(board):
	count = 0
	for i in range(BOARD_LEN):
		for j in range(BOARD_LEN):
			if board[i][j] == '.':
				count+=1

	if count==64:
		return True
	else:
		return False

def getPosition(piece, board):

	for i in range(BOARD_LEN):
		for j in range(BOARD_LEN):
			if board[i][j] == piece:
				return (i,j)



def printBoard(board):
	for i in range(BOARD_LEN):
		for j in range(BOARD_LEN):
			print(board[i][j], end='')
		print()

def checkCheck(board, color, x_k,y_k):
	Check = False
	piece = {}
	if color == 'white':
		piece = {'k': 'k','q': 'q','r': 'r','b': 'b','n':'n'}
	else:
		piece = {'k': 'K','q': 'Q','r': 'R','b': 'B','n':'N'}

	stop = {'u': False,'d': False,'l': False,'r': False, 'u_r': False, 'u_l': False,'d_r': False,'d_l': False}
	dd = 1
	while True:
		count = 0
		for s in stop:
			if stop[s] == True:
				count+=1
		if count ==8:
			break

		coord = {'u': (y_k - dd,x_k) ,'d': (y_k + dd,x_k) ,'l': (y_k ,x_k-dd),'r': (y_k ,x_k+dd), 'u_r': (y_k - dd,x_k+dd), 'u_l': (y_k - dd,x_k-dd),'d_r': (y_k + dd,x_k+dd),'d_l': (y_k + dd,x_k-dd)}

		#VERIFICA DIRECOES (FALTA PEAO  E CAVALO)
		for direction in coord:
			cc = coord[direction]
			if cc[0]<0 or cc[0] >= BOARD_LEN or cc[1]<0 or cc[1] >= BOARD_LEN:
				stop[direction] = True
				continue

			pp = board[cc[0]][cc[1]]
			#if pp =='b':
			#	print(f'{cc[0]} {cc[1]}')

			if stop[direction] == False:
				#se a peca for outro rei 'k':
				if dd==1 and pp==piece['k']:
					Check=True 
					stop[direction] = True
					return Check
				#se a peca q (rainha)
				elif pp==piece['q']:
					Check=True
					stop[direction] = True
					return Check
				#se a peca for r (torre) e tiver na horizontal ou vertical
				elif pp == piece['r'] and (cc[0] == y_k or cc[1] == x_k):
					Check=True
					stop[direction] = True
					return Check
				elif pp == piece['b'] and ( (cc[0] == (-cc[1] + y_k + x_k) ) or (cc[0] == (cc[1] + y_k -x_k) )):
					Check=True
					stop[direction] = True
					return Check			
				elif pp != '.':
					stop[direction] = True

		dd+=1

	#verificar peao
	#------------------------------------
	if color == 'white':
		if x_k-1>=0 and y_k -1 >=0 and board[y_k-1][x_k-1]=='p':
			Check=True
			return Check
		if x_k+1< BOARD_LEN and y_k - 1 >=0  and board[y_k-1][x_k+1]=='p':
			Check=True
			return Check
	else:#black
		if x_k-1>=0 and y_k +1 < BOARD_LEN and board[y_k+1][x_k-1]=='P':
			Check=True
			return Check
		if x_k+1< BOARD_LEN and y_k +1 < BOARD_LEN and board[y_k+1][x_k+1]=='P':
			Check=True
			return Check

	#verificar cavalos
	coord = [(y_k-1,x_k+2),(y_k-2,x_k+1), (y_k-2, x_k-1),(y_k-1,x_k-2),(y_k+1,x_k-2),(y_k+2,x_k-1),(y_k+2,x_k+1),(y_k+1,x_k+2)]
	for cc in coord:
		y = cc[0]
		x = cc[1]
		if x>=0 and x< BOARD_LEN and y>=0 and y< BOARD_LEN and board[y][x] == piece['n']:
			Check=True
			return Check
			break	

	return Check



if __name__ == '__main__':

	n_game=0
	while True:
		n_game+=1
		board = []
		for x in range(BOARD_LEN):
			line = stdin.readline()
			board.append(line)

		if EmptyBoard(board):
			break

		stdin.readline() # linha vazia

		#encontrar posicao dos reis
		k_position = getPosition('k', board) #petro
		K_position = getPosition('K', board) #branco
		Check_k = False
		Check_K = False


		#print(f'{k_position} {K_position}')
		#printBoard(board)
		
		#verificar se esta sob ataque (BRANCO)
		x_k = K_position[1]
		y_k = K_position[0]
		Check_White =  checkCheck(board,'white',x_k,y_k)
		x_k = k_position[1]
		y_k = k_position[0]
		Check_Black =  checkCheck(board,'black',x_k,y_k)

		#imprir saida
		if Check_White == True:
			print(f'Game #{n_game}: white king is in check.')
		elif Check_Black == True:
			print(f'Game #{n_game}: black king is in check.')
		else:
			print(f'Game #{n_game}: no king is in check.')			


#@END_OF_SOURCE_CODE*/

File: ID10196_CheckTheCheck.cpp
/* @JUDGE_ID: 10196 C++ "Check the Check"  */

/* @BEGIN_OF_SOURCE_CODE*/
#include <stdio.h>
#include<iostream>
#include <stdlib.h> /*Character handling functions*/
#include<string.h>


using namespace std;

const int INF = 0x3F3F3F3F;
const int NULO = -1;
const double EPS = 1e-10;

#define BOARD_LEN 8
#define WHITE 1
#define BLACK 0
char BOARD[BOARD_LEN][BOARD_LEN];
char MARK_BOARD[BOARD_LEN][BOARD_LEN];
char PIECES[2][6] = {{'p','n','b','r','q','k'},{'P','N','B','R','Q','K'}};

typedef struct
{
	int x;
	int y;
} Point;

void cleanBoard(char board[][BOARD_LEN]){
	for (int i = 0; i < BOARD_LEN; ++i)
	{
		for (int j = 0; j < BOARD_LEN; ++j)
		{
			board[i][j] = '.';
		}
	}

}

int isBoardEmpty(){
	for (int i = 0; i < BOARD_LEN; ++i)
	{
		for (int j = 0; j < BOARD_LEN; ++j)
		{
			if (BOARD[i][j] != '.')
				return 0;
		}
	}

	return 1;
}

void saveLine(char* s, int index){
	for (int j = 0; j < BOARD_LEN; ++j)
	{
		BOARD[index][j] = s[j];
	}
}

void printBoard(char board[][BOARD_LEN]){
	for (int i = 0; i < BOARD_LEN; ++i)
	{
		for (int j = 0; j < BOARD_LEN; ++j)
		{
			printf("%c",board[i][j]);
		}
		printf("\n");
	}

}

int isBlackWhite(char p, int black_white){
	for (int i = 0; i < 6; ++i)
	{
		if (PIECES[black_white][i]==p)
		{
			return 1;
		}
	}
	return 0;
}

void checkRoot(int idx, int jdx, int black_white){
	char p, king;
	int color_king;
	color_king = black_white ? BLACK: WHITE;
	king = PIECES[color_king][5];

	//UP
	for (int i = idx-1; i >= 0; --i){
		p =  BOARD[i][jdx];
		if ( p=='.' || p==king) 
			MARK_BOARD[i][jdx] = '*'; //VERIFICAR ISSO
		else
			break;
	}

	//DOWN
	for (int i = idx+1; i < BOARD_LEN; ++i){
		p =  BOARD[i][jdx];
		if ( p=='.' || p==king) 
			MARK_BOARD[i][jdx] = '*'; //VERIFICAR ISSO
		else
			break;
	}

	//LEFT
	for (int j = jdx-1; j >=0; --j){
		p =  BOARD[idx][j];
		if ( p=='.' || p==king) 
			MARK_BOARD[idx][j] = '*'; //VERIFICAR ISSO
		else
			break;
	}

	//RIGHT
	for (int j = jdx+1; j < BOARD_LEN; ++j){
		p =  BOARD[idx][j];
		if ( p=='.' || p==king) 
			MARK_BOARD[idx][j] = '*'; //VERIFICAR ISSO
		else
			break;
	}
}

void checkBishop(int idx, int jdx, int black_white){
	int i,j,dif;
	char p, king;
	int color_king;
	color_king = black_white ? BLACK: WHITE;
	king = PIECES[color_king][5];

	//up right
	dif = 1;
	while(1){
		i = idx-dif;
		j = jdx+dif;
		if (i>=0  && j< BOARD_LEN){
			p = BOARD[i][j];
			if ( p=='.' || p==king)
				MARK_BOARD[i][j] = '*';
			else
				break;
		}else
			break;
		dif++;
	}

	//up left
	dif = 1;
	while(1){
		i = idx-dif;
		j = jdx-dif;
		if (i>=0  && j>=0){ 
			p = BOARD[i][j];
			if ( p=='.' || p==king)
				MARK_BOARD[i][j] = '*';
			else
				break;
		}else
			break;
		dif++;
	}

	//down left
	dif = 1;
	while(1){
		i = idx+dif;
		j = jdx-dif;
		if (i< BOARD_LEN  && j>=0){
			p = BOARD[i][j];
			if ( p=='.' || p==king)
				MARK_BOARD[i][j] = '*';
			else
				break;
		}else
			break;
		dif++;
	}

	//down right
	dif = 1;
	while(1){
		i = idx+dif;
		j = jdx+dif;
		if (i< BOARD_LEN  && j< BOARD_LEN){ 
			p = BOARD[i][j];
			if ( p=='.' || p==king)
				MARK_BOARD[i][j] = '*';
			else
				break;
		}else
			break;
		dif++;
	}

}

void checkBoard(int black_white){
	char p;

	///
	for (int i = 0; i < BOARD_LEN; ++i)
	{
		for (int j = 0; j < BOARD_LEN; ++j)
		{
			p = BOARD[i][j];
			if (!isBlackWhite(p, black_white))
				continue;

			if( p== PIECES[black_white][0]){ //'p' - 'P'
				if (black_white == WHITE){
						if ( i-1>= 0 && j-1 >= 0 ) MARK_BOARD[i-1][j-1] = '*';
						if ( i-1>= 0 && j+1 < BOARD_LEN ) MARK_BOARD[i-1][j+1] = '*';
				}else{//BLACK
						if ( i+1< BOARD_LEN && j-1 >= 0 ) MARK_BOARD[i+1][j-1] = '*';
						if ( i+1 < BOARD_LEN && j+1 < BOARD_LEN ) MARK_BOARD[i+1][j+1] = '*';
				}
				continue;
			}

			if(p == PIECES[black_white][1]){ //'n' - 'N'
				if ( i-1>= 0 && j-2 >= 0 ) MARK_BOARD[i-1][j-2] = '*';
				if ( i+1 < BOARD_LEN && j-2 >= 0 ) MARK_BOARD[i+1][j-2] = '*';

				if ( i-2>= 0 && j-1 >= 0 ) MARK_BOARD[i-2][j-1] = '*';
				if ( i+2 < BOARD_LEN && j-1 >= 0 ) MARK_BOARD[i+2][j-1] = '*';

				if ( i-2>= 0 && j+1 < BOARD_LEN ) MARK_BOARD[i-2][j+1] = '*';
				if ( i+2 < BOARD_LEN && j+1 < BOARD_LEN ) MARK_BOARD[i+2][j+1] = '*';

				if ( i-1>= 0 && j+2 < BOARD_LEN ) MARK_BOARD[i-1][j+2] = '*';
				if ( i+1 < BOARD_LEN && j+2 < BOARD_LEN ) MARK_BOARD[i+1][j+2] = '*';

				continue;
			}


			if (p==PIECES[black_white][2]){ //'b' -'B'
				checkBishop(i,j,black_white);
				continue;
			}

			if(p==PIECES[black_white][3]){ //'r' -'R'
				checkRoot(i,j,black_white);
				continue;
			}

			if(p==PIECES[black_white][4]){// 'q' - 'Q'
				checkBishop(i,j,black_white);
				checkRoot(i,j,black_white);
				continue;
			}
			if(p==PIECES[black_white][5]){ // 'K' - 'k'
					if ( i-1>= 0 && j-1 >= 0 ) MARK_BOARD[i-1][j-1] = '*';
					if ( i-1>= 0 ) MARK_BOARD[i-1][j] = '*';
					if ( i-1>= 0 && j+1 < BOARD_LEN ) MARK_BOARD[i-1][j+1] = '*';

					if (j-1 >= 0 ) MARK_BOARD[i][j-1] = '*';
					if (j+1 < BOARD_LEN ) MARK_BOARD[i][j+1] = '*';

					if ( i+1<BOARD_LEN  && j-1 >= 0 ) MARK_BOARD[i+1][j-1] = '*';
					if ( i+1<BOARD_LEN ) MARK_BOARD[i+1][j] = '*';
					if ( i+1<BOARD_LEN  && j+1 < BOARD_LEN ) MARK_BOARD[i+1][j+1] = '*';
			}
			
			
		}
	}

}

void getPosition(char p, int* idx, int* jdx){
	for (int i = 0; i < BOARD_LEN; ++i)
	{
		for (int j = 0; j < BOARD_LEN; ++j)
		{
			if (BOARD[i][j] == p){
				*idx = i;
				*jdx = j;
				return;
			}
		}
	}
}

int main() {
	char s[10], cc;
	int idx, jdx;
	int count = 1;
	while(1){
		//ler a entrada:
		for (int i = 0; i < BOARD_LEN; ++i)
		{
			scanf("%s",s);
			//printf("%s\n",s);
			saveLine(s,i);
		}

		if(isBoardEmpty())
			break;

		//verificar ataque preto
		cleanBoard(MARK_BOARD);
		checkBoard(BLACK);
		getPosition('K',&idx,&jdx);
		//cout<<idx<<"  "<<jdx<<endl;
		cc = MARK_BOARD[idx][jdx];

		if (cc=='*'){
			printf("Game #%d: white king is in check.\n", count);
		}else{
			//verifica ataque branco
			cleanBoard(MARK_BOARD);
			checkBoard(WHITE);
			getPosition('k',&idx,&jdx);
			cc = MARK_BOARD[idx][jdx];
			if (cc=='*')
				printf("Game #%d: black king is in check.\n", count);
			else
				printf("Game #%d: no king is in check.\n", count);

		}


		
		count++;

	}

	return 0;
}

/*@END_OF_SOURCE_CODE*/